### PR TITLE
TASK-21: Break Down travel route into possible granular paths

### PR DESCRIPTION
**Pull request by ys-lin**
This PR implements TASK-21 (#72).

Implemented floor to floor indoor navigation.
Implemented building to building indoor navigation.

Pre-requisite for: #43 and #44

### CODE REVIEW

File: src/services/pathfinding.service.ts [hunk 1]
```diff
@@ -1,7 +1,19 @@
...
+  TravelNode,
+  PQItem,
+  POI,
+  Line,
+  POICategory,
+  FloorPath,
+  ConnectorPOI
+} from "../types/main";
+import * as floors from "../constants/floors.data";
+import { POIInfo } from "../constants/poi.data";
```
Comment (JimLau49) [inline src/services/pathfinding.service.ts]: I believe these 2 can be merged as ` import { POIInfo, buildingFloors } from "../constants" `

Comment (bit172) [reply src/services/pathfinding.service.ts]: Is this because of the new index files?

Comment (JimLau49) [reply src/services/pathfinding.service.ts]: Yeah exactly, probably need to change `floors` to `buildingFloors` too

File: __tests__/pathfinding.service.test.js [hunk 2]
```diff
@@ -70,6 +71,106 @@ describe("Find the shortest path on a given floor", () => {
...
+      displayName: "Stairs 1",
+      description: "",
+      location: {
+        latitude: 45.497354,
+        longitude: -73.578713
+      },
+      buildingId: BuildingId.H,
+      level: 3,
+      category: POICategory.Stairs
+    });
```
Comment (r-saba) [inline __tests__/pathfinding.service.test.js]: I find it a little odd that mock data is being pushed into constants. Can this be in `__testData__` or something similar?

File: src/constants/poi.data.ts [hunk 3]
```diff
@@ -1735,19 +1760,6 @@ export const POIInfo: POI[] = [
...
-
-    description: "Exit/Entrance to JMSB and the corner of Police St. and Guy",
-    location: {
-      latitude: 45.495216991097614,
-      longitude: -73.5785603480603
-    },
-    buildingId: BuildingId.MB,
-    level: 1,
-    category: POICategory.Exit
-  },
```
Comment (r-saba) [inline src/constants/poi.data.ts]: Why was this removed?

File: src/services/pathfinding.service.ts [hunk 4]
```diff
@@ -14,20 +26,220 @@ class PathFindingService {
     return PathFindingService.instance;
   }
 
+  /**
+   * Returns a list of paths between two pois.
+   *
+   * These POI's can be within the same building, or in two different buildings.
```
Comment (r-saba) [inline src/services/pathfinding.service.ts]: line 30 uses `pois` line 32 uses `POI's`. I think throughout the project it's POI's or POIs so can you change line 30 accordingly?

File: src/services/pathfinding.service.ts [hunk 5]
```diff
@@ -14,20 +26,221 @@ class PathFindingService {
...
+    ) as ConnectorPOI;
+
+    // TODO: connect these paths to the outdoor path.
+    return this.findBuildingPath(start, startBuildingExit).concat(
+      this.findBuildingPath(end, endBuildingExit)
+    );
+  };
+
+  /**
+   * Returns a list of paths between two pois within the same building
```
Comment (r-saba) [inline src/services/pathfinding.service.ts]: `pois` --> `POI's`

File: src/constants/poi.data.ts (hunk 3 above)
Comment (bit172) [reply src/constants/poi.data.ts]: We agreed to have only 1 exit per building to simplify indoor navigation

File: __tests__/pathfinding.service.test.js (hunk 2 above)
Comment (bit172) [reply __tests__/pathfinding.service.test.js]: We have to modify the actual objects in `src/constants` since the algorithms use it and we didn't want to duplicate such a large object. We could probably move the pushes into a `before` and `after` block and the new data to `__testData__` How does that sound?

**Review by r-saba — CHANGES_REQUESTED** [review]

File: __tests__/pathfinding.service.test.js [hunk 6]
```diff
@@ -1,12 +1,13 @@
 import * as testData from "./__testData__/pathfinding.data";
 import { POIInfo, buildingFloors } from "../src/constants";
-import { BuildingId } from "../src/types/main";
+import * as constants from "../src/constants";
+import { BuildingId, POICategory } from "../src/types/main";
 import PathFindingService from "../src/services/pathfinding.service";
 
-describe("Find the shortest path on a given floor", () => {
+describe("Find the shortest path between two locations/pois", () => {
```
Comment (RGPosadas) [inline __tests__/pathfinding.service.test.js]: Same as Ragith's comment, `POIs` instead of `pois`
Just CTRL+F for lowercase poi and replace them with uppercase POI

File: src/constants/poi.data.ts [hunk 7]
```diff
@@ -64,6 +64,18 @@ export const POIInfo: POI[] = [
     level: 8,
     category: POICategory.EscalatorDown,
   },
+  {
+    id: "5454457b-debf-4511-9bd0-6bc61c9177a3",
+    displayName: "  Exit",
```
Comment (RGPosadas) [inline src/constants/poi.data.ts]: Why the whitespace in `"  Exit"`?

File: src/services/pathfinding.service.ts [hunk 8]
```diff
@@ -14,20 +25,221 @@ class PathFindingService {
...
+    });
+
+    return pathsToConnectors;
+  };
+
+  /**
+   * Searches through valid paths between start and end POI to find the shortest one.
+   *
+   * @param startFloorPathsToConnectors
+   * @param endFloorPathsToConnectors
```
Comment (RGPosadas) [inline src/services/pathfinding.service.ts]:
```
   * @param startFloorPathsToConnectors
   * @param endFloorPathsToConnectors
```
These may need more descriptions
Might also just double check for other params in the file that don't have descriptions on them, especially since this file is quite huge

**Review by RGPosadas — CHANGES_REQUESTED** [review]
So far my small review comments are non-blocking.

This is what's BLOCKING:
It's the same thing as I mentioned in this [PR](https://github.com/RGPosadas/WayFinder/pull/67#issuecomment-604033237). As someone who didn't work on this ticket, how do we debug/play around with the implemented service? Where are the instructions?

Please write debugging instructions down (maybe it's better if we add it in the Wiki page?).

Otherwise, nice work guys!

**Comment by cristian-aldea** [comment]
@RGPosadas 
> Please write debugging instructions down (maybe it's better if we add it to the Wiki page?).

Good point, but I feel like that kind of visual testing/debugging would be best done for the user story that this task support (#43). Writing that debug code would basically be writing whatever code we need for #43, which would bloat this ticket more than it needs to.

For now, we could provide screenshots to show how the path would generally look, but I think the reviewers should just base their acceptance criteria for this task on if the unit tests are passing for now.

File: __tests__/pathfinding.service.test.js (hunk 2 above)
Comment (r-saba) [reply __tests__/pathfinding.service.test.js]: Maybe clone the data in your test file?

**Comment by bit172** [comment]
Here is a screenshot of one the test paths from H961-17 to the MB Men's Washroom:
Screenshot_20200408-165310_1_20Screenshot_20200408-165409_20

Here is another screenshot from the H961-17 to H8 Women's Bathroom 
Screenshot_20200408-165310_1_20Screenshot_20200408-171713_1_20

I did this by modifying `src/components/map-overlays/travel-node-debug.component.tsx` so that it could display the paths. I will not show implementation details as it is very manual and would essentially be the implementation of #43.

File: __tests__/pathfinding.service.test.js (hunk 2 above)
Comment (bit172) [reply __tests__/pathfinding.service.test.js]: We came up with something. Check out the new code since this file is outdated. Let me know what you think

File: src/constants/poi.data.ts (hunk 7 above)
Comment (bit172) [reply src/constants/poi.data.ts]: Seems to be a mistake, will fix it

File: src/services/pathfinding.service.ts (hunk 8 above)
Comment (bit172) [reply src/services/pathfinding.service.ts]: Yes, I will add descriptions to the parameters of the functions in this file.

File: __tests__/pathfinding.service.test.js [hunk 9]
```diff
@@ -75,56 +75,24 @@ describe("Find the shortest path on a given floor", () => {
...
-      description: "",
-      location: {
-        latitude: 45.497354,
-        longitude: -73.578713
-      },
-      buildingId: BuildingId.H,
-      level: 3,
-      category: POICategory.Stairs
-    });
+    constants.buildingFloors = buildingFloors.push(testData.h3rdFloorPoiMock);
```
Comment (r-saba) [inline __tests__/pathfinding.service.test.js:78]: I like it! Also very easy to read what's happening

File: __tests__/pathfinding.service.test.js (hunk 2 above)
Comment (r-saba) [reply __tests__/pathfinding.service.test.js]: Nice, looks resolved to me!

**Review by r-saba — DISMISSED** [review]
LGTM!

#### Acceptance Criteria
- [x] Create a service that takes in a start and end location that the user has selected
- [x] The service contains an algorithm to do what is described in the point above
- [x] The algorithm uses information about POIs to break down the route into paths
- [x] Algorithm gives back a list of viable paths

#### Notes
My comments were addressed. Screenshots look good. Good work!!! Mentioned some minor which can be addressed in the actual user story.

**Review by RGPosadas — DISMISSED** [review]
#### Detox Tests
All tests pass.

#### Acceptance Criteria
- [x] Create a service that takes in a start and end location that the user has selected
- [x] The service contains an algorithm to do what is described in the point above
- [x] The algorithm uses information about POIs to break down the route into paths
- [x] Algorithm gives back a list of viable paths

#### Notes
Thank you for addressing my comments! The screenshots are a great addition - it was a bit hard to imagine the paths otherwise. Great work everyone!

#### Before Merging
I see some funky commits such as:
image

Please rebase and add more descriptive commits under the header commit, like this example:
image

**Review by RGPosadas — APPROVED** [review]
Thanks for the rebase! Re-approving.
